src/metabase/models/app/graph.clj

```diff
@@ -0,0 +1,72 @@
+(ns metabase.models.app.graph
+  "Code for generating and updating the App permissions graph. App permissions
+  are based on the permissions of the app's collection."
+  (:require [clojure.data :as data]
+            [metabase.models.collection-permission-graph-revision :as c-perm-revision
+             :refer [CollectionPermissionGraphRevision]]
+            [metabase.models.collection.graph :as graph]
+            [metabase.models.permissions :as perms]
+            [metabase.models.permissions-group :as perms-group]
+            [metabase.models.setting :as setting :refer [defsetting]]
+            [metabase.public-settings.premium-features :as premium-features]
+            [metabase.util.i18n :as i18n :refer [tru]]
+            [metabase.util.schema :as su]
+            [schema.core :as s]
+            [toucan.db :as db]))
+
+(def ^:private AppPermissions
+  graph/CollectionPermissions)
+
+(def ^:private GlobalPermissionsGraph
+  {:revision s/Int
+   :groups   {su/IntGreaterThanZero AppPermissions}})
+
+(s/defn ^:private set-all-users-app-permission
+  [permission :- AppPermissions]
+  (setting/set-value-of-type! :keyword :all-users-app-permission permission))
+
+(defsetting all-users-app-permission
+  "App permission of the All Users group"
+  :type :keyword
+  :visibility :internal
+  :default :none
+  :setter (fn [v]
+            (set-all-users-app-permission (cond-> v (string? v) keyword))))
+
+(defn set-default-permissions!
+  "Sets the default permissions for the ''All Users'' group on`app` as specified
+  by `all-users-app-permission` if advanced permissions are not available."
+  [app]
+  (when-not (premium-features/has-feature? :advanced-permissions)
+    (graph/update-collection-permissions! (:id (perms-group/all-users))
+                                          (:collection_id app)
+                                          (all-users-app-permission))))
```

Comment:
This feels off to me but I'm not sure if there's a better way.

My understanding of is that one principle of our permission system is that it shouldn't require these kind of write-time updates.

If I understand this correctly, the problem that the use of this in `api.app` is trying to solve is that apps do not have a proper root to inherit permissions from for new apps.

Would it be (a) possible and (b) better instead in oss to either (I don't know)
1. generate the global graph based on `all-users-app-permission`
2. check `all-users-app-permission` directly and skip the graph check in oss
3. check `all-users-app-permission` directly and app permissions are still based on the `collection` graph. (so in oss, app access would be gated on root-collection permission)

Reply:
I agree that this is ugly. This is an attempt to reuse collection permissions for apps as we decided in the pod. My understanding is that the whole point of reusing collections permissions is to avoid implementing separate checks.
@camsaul WDYT, is reusing the collection permissions (at least in the beginning) a good idea? If it is, do you see a better way of reusing them? If it is not, what way would you implement this?